### PR TITLE
Example/h4alignment handbook

### PR DESCRIPTION
**Pull request by deep-diver**

### CODE REVIEW

File: examples/fine-tuning/alignment-handbook/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+You can follow the same steps to run fine-tuning job on dstack. [`train,dstack.yaml`](./train.dstack.yaml) shows a concrete example:
+
+```yaml
+type: task
+
+python: "3.11"
+
+env:
+  - HUGGING_FACE_HUB_TOKEN=<YOUR-HF-ACCESS-TOKEN>
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/README.md]: Better:

```
env:
  - HUGGING_FACE_HUB_TOKEN
```

File: examples/fine-tuning/alignment-handbook/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  - python -m pip install flash-attn --no-build-isolation
+
+  - pip install wandb
+  - wandb login $WANDB_API_KEY
+
+  - | 
+    ACCELERATE_LOG_LEVEL=info 
+    accelerate launch 
+    --config_file recipes/accelerate_configs/multi_gpu.yaml 
+    --num_processes=$NUM_GPUS
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/README.md]: The environment variables `CONFIG_PATH` and `NUM_GPUS` won't be passed properly if you don't list then in `env` (see `HUGGING_FACE_HUB_TOKEN` above)

File: examples/fine-tuning/alignment-handbook/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+- `HUGGING_FACE_HUB_TOKEN` should be set as environment variable so that the fine-tuned model and its checkpoints could be pushed to the Hugging Face Hub (for downloading gated models and private datasets as well).
+- the first 4 lines clone `alignment-handbook` repo, copy local `config.yaml` into the repository.
+- the second 3 lines install dependencies required to use `alignment-handbook`.
+- the third 2 lines login to Weights and Biases. These line should be present if you have listed `wandb` under `report_to` field in `config.yaml`.
+- the last line launch the SFT job on 2 GPUs.
+
+Now, simply run the `dstack run` command as below:
+
+```consnole
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/README.md]: typo, must be `console`

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+type: task
+
+python: "3.11"
+
+env:
+  - CONFIG_PATH=configs/sample_config.yaml
+  - NUM_ACCELERATORS=<Number of accelerators specified>
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/train.dstack.yaml]: Lets replace `<Number of accelerators specified>` with `2`

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  - HUGGING_FACE_HUB_TOKEN=<Your Hugging Face access token>
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/train.dstack.yaml]: Better:
```
env:
  - HUGGING_FACE_HUB_TOKEN
 ```

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  - cp "$CONFIG_PATH" alignment-handbook/recipes/custom/config.yaml
+
+  - cd alignment-handbook
+  - python -m pip install .
+  - python -m pip install flash-attn --no-build-isolation
+
+  - pip install wandb
+  - wandb login $WANDB_API_KEY
+
+  - ACCELERATE_LOG_LEVEL=info accelerate launch --config_file recipes/accelerate_configs/multi_gpu.yaml --num_processes=$NUM_ACCELERATORS scripts/run_sft.py recipes/custom/config.yaml
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/train.dstack.yaml]: In `README.md`, you use `GPU_NUM` and here `NUM_ACCELERATORS`. We could have the same names.

File: docs/examples/infinity.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
+# Infinity
```
Comment (peterschmidt85) [inline docs/examples/infinity.md]: I suggest that we mve Infinity to a separate PR

File: examples/fine-tuning/alignment-handbook/README.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+$ CONFIG_PATH=config.yaml \
+NUM_GPUS=2 \
+WANDB_API_KEY=<YOUR-W&B-API-KEY> \
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/README.md:89]: Add `HUGGING_FACE_HUB_TOKEN=<YOUR-HF-TOKEN> \`

**Review by peterschmidt85 — CHANGES_REQUESTED** [review]

**Comment by deep-diver** [comment]
@peterschmidt85 

addressed your comment. please take a look one more time! Thanks!

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml [hunk 9]
```diff
@@ -0,0 +1,39 @@
+type: task
+
+python: "3.11"
+
+env:
+  - HUGGING_FACE_HUB_TOKEN
+  - CONFIG_PATH
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/train.dstack.yaml]: Do we need it as an environment variable? I suppose it is always `config.yml`, right? IMO to simplify, I'd hardcode it.

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  - NUM_GPUS
```
Comment (peterschmidt85) [inline examples/fine-tuning/alignment-handbook/train.dstack.yaml]: In 0.18.1, we will add a system environment variable `DSTACK_GPUS` that will be automatically assigned buy dstack. So for now, I'd suggest that we hardcode it to `2` and do not introduce a user environment variable.

**Review by peterschmidt85 — CHANGES_REQUESTED** [review]
Left a few more commens.

File: examples/fine-tuning/alignment-handbook/train.dstack.yaml (hunk 9 above)
Comment (deep-diver) [reply examples/fine-tuning/alignment-handbook/train.dstack.yaml]: For the sake of simplicity of this as a tutorial, I removed `CONFIG_PATH` and `NUM_GPU` environment variables. However, I do think it is common not to hardcode `yaml` file path. It is very common to have lots of YAMLs with different configurations to run multiple experiments with different hyper-params.

**Comment by deep-diver** [comment]
@peterschmidt85 

all addressed!

**Review by peterschmidt85 — APPROVED** [review]

**Pull request merged**